### PR TITLE
Sponsor service

### PR DESCRIPTION
**Pull request by egnwd**
⚠️ _This PR is huge_

This PR allows a sponsor to login with email-password and view a list of  CVs and download them individually.

I'm throwing this up here so I can look and review myself, feel free to review yourself.

### CODE REVIEW

File: auth/auth.go [hunk 1]
```diff
@@ -1,87 +1,103 @@
...
 
+type PasswordAuth = auth
+
+var scopes = []string{oauthService.UserinfoEmailScope, oauthService.UserinfoProfileScope}
+
 const (
 	login    = "/login"
 	logout   = "/logout"
 	callback = "/callback"
 	token    = "/jwt/token"
```
Comment (egnwd) [inline auth/auth.go]: Can be removed

File: auth/jwt.go [hunk 2]
```diff
@@ -1,65 +1,175 @@
...
+		if onetime {
+			session, err := cookieJar.Get(r, sessionKey)
+			if err != nil {
+				http.Error(w, err.Error(), http.StatusInternalServerError)
+				return
 			}
 
-			return auth.jwt.secret, nil
-		})
+			jti, err := generateAndStore("jti", session, w, r)
```
Comment (egnwd) [inline auth/jwt.go]: `const jtiKey = "jti"`

File: auth/jwt.go [hunk 3]
```diff
@@ -1,65 +1,175 @@
...
+		}
+
+		log.Println("Checking claims")
+		if claims, ok := token.Claims.(*Claims); ok {
+			if !roles.HasRole(r, &claims.UserIdentifier, validRoles...) {
+				http.Error(w, "Invalid access", http.StatusUnauthorized)
+				return
+			}
+
+			log.Println("Valid Roles")
```
Comment (egnwd) [inline auth/jwt.go]: remove

File: auth/jwt.go [hunk 4]
```diff
@@ -1,65 +1,175 @@
...
+
+			if err = checkJti(claims, onetime, auth, w, r); err != nil {
+				http.Error(w, "Invalid access", http.StatusUnauthorized)
+				return
+			}
+
+			log.Println("Valid JTI")
```
Comment (egnwd) [inline auth/jwt.go]: remove

File: auth/jwt.go [hunk 5]
```diff
@@ -1,65 +1,175 @@
...
 	})
 }
+
+func extractJti(auth Auth, w http.ResponseWriter, r *http.Request) (string, error) {
+	session, err := auth.session(r, sessionKey)
+	if err != nil {
+		return "", err
+	}
+
+	return getAndDelete("jti", session, w, r)
```
Comment (egnwd) [inline auth/jwt.go]: `jtiKey`

File: auth/user.go [hunk 6]
```diff
@@ -28,13 +36,13 @@ func (auth *Auth) getCurrentUser(r *http.Request) (UserIdentifier, error) {
 	return nil, nil
 }
 
-func (auth *Auth) setCurrentUser(w http.ResponseWriter, r *http.Request, userID UserIdentifier) error {
-	session, err := auth.store.Get(r, sessionKey)
+func setCurrentUser(auth Auth, w http.ResponseWriter, r *http.Request, userID *UserIdentifier) error {
+	session, err := auth.session(r, sessionKey)
```
Comment (egnwd) [inline auth/user.go]: use `cookieJar`

File: auth/user.go [hunk 7]
```diff
@@ -44,8 +52,8 @@ func (auth *Auth) setCurrentUser(w http.ResponseWriter, r *http.Request, userID
 	return nil
 }
 
-func (auth *Auth) deleteCurrentUser(w http.ResponseWriter, r *http.Request) error {
-	session, err := auth.store.Get(r, sessionKey)
+func deleteCurrentUser(auth Auth, w http.ResponseWriter, r *http.Request) error {
+	session, err := auth.session(r, sessionKey)
```
Comment (egnwd) [inline auth/user.go]: use `cookieJar`

File: client/dist/sponsors.html [hunk 8]
```diff
@@ -0,0 +1,14 @@
+<!DOCTYPE html>
```
Comment (egnwd) [inline client/dist/sponsors.html]: remove

File: client/dist/students.html [hunk 9]
```diff
@@ -8,6 +8,6 @@
   </head>
```
Comment (egnwd) [inline client/dist/students.html]: remove

File: client/src/fetch.js [hunk 10]
```diff
@@ -1,21 +1,17 @@
-import request from 'superagent';
+import request from 'request-promise-native';
```
Comment (egnwd) [inline client/src/fetch.js]: double check if superagent can be used

File: client/src/views/Home/index.jsx [hunk 11]
```diff
@@ -1,3 +1,3 @@
 import HomeView from './HomeView';
 
-export default HomeView;
+export default HomeView
```
Comment (egnwd) [inline client/src/views/Home/index.jsx]: add this back 🙄

File: client/src/views/Login/Login.jsx [hunk 12]
```diff
@@ -0,0 +1,72 @@
+import React from 'react';
+import Header from 'Components/Header';
+import md5 from 'md5';
+import request from 'request-promise-native';
```
Comment (egnwd) [inline client/src/views/Login/Login.jsx]: double check superagent

File: client/src/views/Login/Login.jsx [hunk 13]
```diff
@@ -0,0 +1,72 @@
...
+      this.setState({error: "The email and/or password you entered was incorrect"});
+    }
+  }
+
+  handleChange = (event) => this.setState({ [event.target.name]: event.target.value })
+
+  render() {
+    let {redirect, error, email, password, nextPathname} = this.state;
+    if (!!redirect) {
+      return (<Redirect to={nextPathname || redirect} />);
```
Comment (egnwd) [inline client/src/views/Login/Login.jsx]: is `nextPathname` needed?

File: client/src/views/Students/StudentProfile.jsx [hunk 14]
```diff
@@ -54,9 +53,11 @@ export default class StudentProfile extends React.Component {
         on('progress', event => {
           progress(event.percent);
         });
+
       this.setState({upload: false})
     } catch (e) {
       console.log(e)
+      console.log("inspect", fs)
```
Comment (egnwd) [inline client/src/views/Students/StudentProfile.jsx]: remove

**Review by jamesprinc3 — APPROVED** [review]

**Pull request merged**